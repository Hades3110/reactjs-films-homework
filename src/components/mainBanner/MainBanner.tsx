import React, { useState, useEffect } from 'react'
import { FilmInfoInterface, GenreInterface, getFilmVideo, getHeadFilm } from '../../services/filmService'
import StarRatings from 'react-star-ratings'
import styles from './mainBanner.module.scss'
import ReactPlayer from 'react-player'

const initialState: FilmInfoInterface = {
    id: 0,
    backdrop_path: '',
    page: 1,
    genres: [],
    title: '',
    results: [],
    total_pages: 1,
    total_results: 1,
    vote_average: 1,
    runtime: 1,
    overview: ''
}

const MainBanner: React.FC = () => {

    const [filmInfo, setFilmInfo] = useState<FilmInfoInterface>(initialState)
    const [overview, setOverview] = useState<boolean>(false)
    const [watch, setWatch] = useState<boolean>(false)
    const [videoKey, setVideoKey] = useState<string>('')

    useEffect(() => {
        getHeadFilm().then(res => setFilmInfo(res))
        if (filmInfo.id) getFilmVideo(filmInfo.id).then(res => setVideoKey(res.results[0].key))
    }, [filmInfo.id])

    const bgImage: string = filmInfo.backdrop_path
    const vote = filmInfo.vote_average ? +(filmInfo.vote_average / 2).toFixed(1) : 0
    const runtime = `${~~(filmInfo.runtime / 60)}h  ${filmInfo.runtime % 60}m`

    return (<>
        {watch ? <div className={styles.video} onClick={() => {
            setWatch(false)
        }
        }>
            <ReactPlayer
                className={styles.player}
                url={`https://www.youtube.com/watch?v=${videoKey}`}
                controls={true}
                width={'60%'}
                height={'70vh'}
            />
        </div> : ''}
        <div className={styles.majorFilm} style={{
            backgroundImage: bgImage ? `url('https://image.tmdb.org/t/p/original${bgImage}')` : ''
        }}>
            <div className={styles.infoAndRating}>
                <div className={styles.infoAndRating__moreInfo}>
                    <div>
                        <h1>{filmInfo.title}</h1>
                        <ul>
                            {filmInfo.genres ? filmInfo.genres.map((el: GenreInterface) => <li key={el.id}>{el.name}</li>) : ''}
                            <li>|</li>
                            <li>{runtime}</li>
                        </ul>
                    </div>
                    <div className={styles.overview} style={overview ? { display: 'block' } : { display: 'none' }}>{filmInfo.overview}</div>
                </div>
                <div className={styles.infoAndRating__ratingAndBtn}>
                    <div className={styles.infoAndRating__ratingAndBtn__rating}>
                        <StarRatings
                            rating={vote}
                            starRatedColor='#5db9e5'
                            numberOfStars={5}
                            name='rating'
                            starDimension={'26px'}
                        />
                        <div className={styles.infoAndRating__ratingAndBtn__rating__num}>{vote}</div>
                    </div>
                    <div className={styles.infoAndRating__ratingAndBtn__btns}>
                        <button className={styles.infoAndRating__ratingAndBtn__btns__watch} onClick={() => {
                            setWatch(true)
                        }
                        }>Watch now</button>
                        <button id={overview ? styles.btnView : ''} onClick={() => setOverview(!overview)}>View Info</button>
                    </div>
                </div>
            </div>
        </div>
    </>)
}

export default MainBanner
