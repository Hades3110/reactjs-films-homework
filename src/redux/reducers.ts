import { combineReducers } from "redux";
import filmListPageCountReducer from "./filmListCountAndCategories/reducer";
import filmListSortReducer from "./filmListSort/reducer";
import mainBannerReducer from "./mainBanner/reducer";

const rootReducer = combineReducers({
    pageCounter: filmListPageCountReducer,
    sortType: filmListSortReducer,
    mainBanner: mainBannerReducer
})

export default rootReducer
