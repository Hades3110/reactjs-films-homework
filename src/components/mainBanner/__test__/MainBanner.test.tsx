import { create } from "react-test-renderer";
import MainBanner from '../MainBanner';
import { createStore } from 'redux';
import rootReducer from '../../../redux/reducers';
import { Provider } from 'react-redux';

describe('MainBanner Component', () => {

  const store = createStore(rootReducer)

  it('MainBanner snapshot test', () => {
    const component = create(<Provider store={store}><MainBanner id={1}/></Provider>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});