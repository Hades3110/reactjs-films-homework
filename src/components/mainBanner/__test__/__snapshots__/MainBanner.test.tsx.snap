// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainBanner Component MainBanner snapshot test 1`] = `
<img
  className="spinner"
  src="test-file-stub"
/>
`;
